contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -190,115 +190,102 @@ class async_protocol_handler
 
   struct invoke_response_handler_base
   {
-    virtual bool handle(int res, const epee::span<const uint8_t> buff, connection_context& context)=0;
-    virtual bool is_timer_started() const=0;
-    virtual void cancel()=0;
-    virtual bool cancel_timer()=0;
-    virtual void reset_timer()=0;
-  };
-  template <class callback_t>
-  struct anvoke_handler: invoke_response_handler_base
-  {
-    anvoke_handler(const callback_t& cb, uint64_t timeout,  async_protocol_handler& con, int command)
-      :m_cb(cb), m_timeout(timeout), m_con(con), m_timer(con.m_pservice_endpoint->get_io_service()), m_timer_started(false),
-      m_cancel_timer_called(false), m_timer_cancelled(false), m_command(command)
-    {
-      if(m_con.start_outer_call())
-      {
-        MDEBUG(con.get_context_ref() << "anvoke_handler, timeout: " << timeout);
-        m_timer.expires_from_now(boost::posix_time::milliseconds(timeout));
-        m_timer.async_wait([&con, command, cb, timeout](const boost::system::error_code& ec)
-        {
-          if(ec == boost::asio::error::operation_aborted)
-            return;
-          MINFO(con.get_context_ref() << "Timeout on invoke operation happened, command: " << command << " timeout: " << timeout);
-          epee::span<const uint8_t> fake;
-          cb(LEVIN_ERROR_CONNECTION_TIMEDOUT, fake, con.get_context_ref());
-          con.close();
-          con.finish_outer_call();
-        });
-        m_timer_started = true;
-      }
-    }
-    virtual ~anvoke_handler()
-    {}
-    callback_t m_cb;
-    async_protocol_handler& m_con;
+  protected:
     boost::asio::deadline_timer m_timer;
-    bool m_timer_started;
-    bool m_cancel_timer_called;
-    bool m_timer_cancelled;
-    uint64_t m_timeout;
-    int m_command;
-    virtual bool handle(int res, const epee::span<const uint8_t> buff, typename async_protocol_handler::connection_context& context)
+    async_protocol_handler& m_con;
```

Comment:
What is the rationale for moving `anvoke_handler` logic into `invoke_response_handler_base` but not combining the two?

Reply:
You mean, why not drop `anvoke_handler` completely in favor of `std::function` type erasure? The only reason is that this saves one allocation (most of the callbacks exceed the small size optimization of `std::` or `boost::` `function`.

If you meant something else, then you'll need to clarify what you meant here.

Reply:
I mean that `invoke_response_handler_base` use to be a pure virtual base class, but now it contains a partial implementation. I'm not saying it's necessarily worse, but now there isn't that clean interface / implementation distinction. I was just wondering the reason for that decision. Since it's not used as a base class anywhere else, why not just combine `invoke_response_handler_base` and `anvoke_handler`?

Reply:
I wanted to remove the `virtual` call for resetting the timer, thus the split of internal members. But otherwise, that wasn't strictly necessary.

However, some indirect call (`std::function`, function pointer, `virtual`, etc.) was still needed because of the templated callback. The `std::vector` has to store multiple different lambdas from `src/p2p/p2p.inl`. So that template argument isn't "fixed" to the levin handler, but rather generic per async invoke.